lib/network.go

```diff
@@ -2139,6 +2290,84 @@ func DecodeHeaderVersion1(rr io.Reader) (*MsgDeSoHeader, error) {
 	return retHeader, nil
 }
 
+func DecodeHeaderVersion2(rr io.Reader) (*MsgDeSoHeader, error) {
+	retHeader := NewMessage(MsgTypeHeader).(*MsgDeSoHeader)
+
+	// PrevBlockHash
+	_, err := io.ReadFull(rr, retHeader.PrevBlockHash[:])
+	if err != nil {
+		return nil, errors.Wrapf(err, "MsgDeSoHeader.FromBytes: Problem decoding PrevBlockHash")
+	}
+
+	// TransactionMerkleRoot
+	_, err = io.ReadFull(rr, retHeader.TransactionMerkleRoot[:])
+	if err != nil {
+		return nil, errors.Wrapf(err, "MsgDeSoHeader.FromBytes: Problem decoding TransactionMerkleRoot")
+	}
+
+	// TstampSecs
+	retHeader.TstampSecs, err = ReadUvarint(rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "MsgDeSoHeader.FromBytes: Problem decoding TstampSecs")
+	}
+
+	// Height
+	retHeader.Height, err = ReadUvarint(rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "MsgDeSoHeader.FromBytes: Problem decoding Height")
+	}
+
+	// The Nonce and ExtraNonce fields are unused in version 2. We skip them
+	// during both encoding and decoding.
+	retHeader.Nonce = 0
+	retHeader.ExtraNonce = 0
+
+	// ProposerPublicKey
+	retHeader.ProposerPublicKey, err = ReadPublicKey(rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "MsgDeSoHeader.FromBytes: Problem decoding ProposerPublicKey")
+	}
+
+	// ProposerVotingPublicKey
+	retHeader.ProposerVotingPublicKey, err = DecodeBLSPublicKey(rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "MsgDeSoHeader.FromBytes: Problem decoding ProposerVotingPublicKey")
+	}
+
+	// ProposedInView
+	retHeader.ProposedInView, err = ReadUvarint(rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "MsgDeSoHeader.FromBytes: Problem decoding ProposedInView")
+	}
+
+	// ValidatorsVoteQC
+	retHeader.ValidatorsVoteQC, err = DecodeQuorumCertificate(rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "MsgDeSoHeader.FromBytes: Problem decoding ValidatorsVoteQC")
+	}
+
+	// ValidatorsTimeoutAggregateQC
+	retHeader.ValidatorsTimeoutAggregateQC = &TimeoutAggregateQuorumCertificate{}
+	if err = retHeader.ValidatorsTimeoutAggregateQC.FromBytes(rr); err != nil {
+		return nil, errors.Wrapf(err, "MsgDeSoHeader.FromBytes: Problem decoding ValidatorsTimeoutAggregateQC")
+	}
+
+	// ProposerVotePartialSignature: we decode the signature if it's present in the byte encoding.
+	// If it's not present, then we set the signature to nil.
+	proposerSignatureBytes, err := DecodeByteArray(rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "MsgDeSoHeader.FromBytes: Problem decoding ProposerVotePartialSignature")
+	}
+	if len(proposerSignatureBytes) != 0 {
+		retHeader.ProposerVotePartialSignature, err = (&bls.Signature{}).FromBytes(proposerSignatureBytes)
+		if err != nil {
+			return nil, errors.Wrapf(err, "MsgDeSoHeader.FromBytes: Problem decoding ProposerVotePartialSignature")
+		}
+	}
```

Comment:
Need to encode/decode the new fields here.

Reply:
Added a new `DecodeOptionalBLSSignature(rr)` function for this too